### PR TITLE
[JSON]Speed up Player and Playlist operations 

### PR DESCRIPTION
**Pull request by DaveTBlake**
Fix API `Player` and `Playlist` operations using post rather than send for application messages that don't need to wait. Minor patch bump to indicate internal change to API . 

In continued pursuit of why the party mode is so very slow to start in certain situations https://github.com/xbmc/xbmc/issues/15130 have found some small way to speed things up.

### CODE REVIEW

**Comment by a1rwulf** [comment]
I'd merely label it with "Improvement" instead of "Fix", as actually nothing is broken.
The change is okish cause we ignored the `SendMsg's` return value anyway.
But not sure if post really fit's for all of those functions.
For example `CPlayerOperations::Open` could actually forward the return value of `SendMsg` to indicate if the operation even worked.
Not sure how the change to `PostMsg` changes this and how json clients now that their play request just failed.

**Comment by DaveTBlake** [comment]
@a1rwulf thanks for interest.
While this did not resolve the slowness of v18 issue I was investigation, this change (along with https://github.com/xbmc/xbmc/pull/15341 ) does **fix** a long standing issue with playlists being unexpectedly slow to load via JSON. Something working unnecessarily slowly counts as a bug in my book, and using 'SendMsg' could easily have been a coding mistake.

All JSON results are indended to indicate that the request was received and valid, they have nothing to do with the indicating the operation was successful. Player operations are asynchronous, requesting playback is one thing, you then need to make another call to discover if playback has successfully started or listen for a notification indicating it has.

**Comment by a1rwulf** [comment]
Alright.

**Review by a1rwulf — APPROVED** [review]

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
I suggest that you postpone all #15130 related attempts to improve things until you can prove that you found the root cause.
You label this as a fix but this is all but a fix.

File: xbmc/interfaces/json-rpc/PlayerOperations.cpp [hunk 1]
```diff
@@ -536,7 +536,7 @@ JSONRPC_STATUS CPlayerOperations::Open(const std::string &method, ITransportLaye
     {
       case PLAYLIST_MUSIC:
       case PLAYLIST_VIDEO:
-        CApplicationMessenger::GetInstance().SendMsg(TMSG_MEDIA_PLAY, playlistid, playlistStartPosition);
+        CApplicationMessenger::GetInstance().PostMsg(TMSG_MEDIA_PLAY, playlistid, playlistStartPosition);
```
[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: Playback requests from both Python and uPNP uses `PostMsg(TMSG_MEDIA_PLAY...` and yet JSON API requests must not. Why?

**Comment by DaveTBlake** [comment]
Investigating #15130 revealed this misuse of SendMsg, but is not the sole reason for this PR. Adding items to the current playlist via JSON has been reported slow for many years, waiting is the cause. But no one cares about JSON or how slow it is etc., prefering to always blame the API client instead.  I was very pleased to find the root cause of that issue, and see it released soonest.

It makes no sense that Python uses PostMsg, as does uPNP, to request playback but that JSON should not do the same. 

Surely notification that the playlist has changed is something that the playlist manager should handle, not be left to the JSON request?

File: xbmc/interfaces/json-rpc/PlayerOperations.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: Python and uPNP is not relevant here. That does not change the fact that your intended code is wrong:

```
        CApplicationMessenger::GetInstance().PostMsg(TMSG_MEDIA_PLAY, playlistid, playlistStartPosition); 
        OnPlaylistChanged();
```

You can't trigger OnPlaylistChanged until this has not happened/completed. The action triggered by OnPlaylistChange might be executed before the action of TMSG_MEDIA_PLAY. Two different messaging mechanisms are involved here.

[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: So clearly the correct fix is to remove OnPlaylistChange() as well?  Then jsut like Python and uPNP something else will handle it.

[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: > So clearly the correct fix is to remove OnPlaylistChange() as well? 

certainly not. there may be clients that rely on this synchronous behaviour of this API.

[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: >there may be clients that rely on this synchronous behaviour of this API.

I am certain that there are not. As you well know player start is asynchronous and so totally unrelated to any possible wait at that point. Before v18 API consumers needed to wait for the  onPlayStart announcement to know playback had started, now that is replaced by onAvStart. API consumers do playlist handling via messaging and messaging is a single queue FIFO. The wait is just a needless delay that prevents the next JSON request from being processed.

[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: In fact the opposite is true. API clients have been mystified why adding items to the queue is so slow, and will be delighted to have playlist requests processed without delays.

[the remark this answers is not in the document]
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: @FernetMenta I am certain that JSON **Player.Open** request should not be notifying the GUI of anything. After all it is the app that defines the current item and tells the GUI if that changes. 

OnPlaylistChanged() notification here is totally wrong

**Comment by DaveTBlake** [comment]
While JSON API may _request_ changes to the current playlist, it should not be responsible for notifying the GUI when the playlist has changed. All notification should be done by the app process or playlist manager.

This had become a mess, with some operations (**swap**, **remove**, **clear**, **Goto**) having `GUI_MSG_PLAYLIST_CHANGED` messages sent by both PlaylistPlayer and JSON. Similar duplication happened with **Player.Open** - the app tells the GUI when the current item has changed.

So I have  extended the changes in this PR, adding `GUI_MSG_PLAYLIST_CHANGED` messages to PlaylistPlayer SetShiffle, SetRepeat, Add and Insert methods, makes then consistent with the others and does it in the right place.  All GUI notification is therefore removed from JSON.

Whether this is called a fix or not I very much believe it should go into v18.1 release. I have fully tested it, and it makes a substantial difference to the speed of JSON playlist operations.

**Comment by DaveTBlake** [comment]
jenkins build this please

File: xbmc/interfaces/json-rpc/PlayerOperations.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (notspiff) [reply xbmc/interfaces/json-rpc/PlayerOperations.cpp]: while fernetmenta is correct in pointing out the change in behaviour, if json clients relies on gui updates, they are relying on a bug. a json command is used to post instructions, the return code is not supposed to carry any information about execution result, just whether or not a command was posted. it's not meant to be a sync api. i consider this a bug fix, with a usability fix to cherry it off.

**Comment by MartijnKaijser** [comment]
jenkins build and merge

**Pull request merged**